### PR TITLE
Fix param_key when model is using relative model naming in base classes

### PR DESCRIPTION
**Pull request by BenMorganIO**
**TL;DR**: Models that inherit from an `Example::Base` instead of a module do not have their namespace removed in their `param_key` or `route_key`.

---

When developing, I've discovered that I opt for using abstract classes instead of modules to contain code. This same style is done with Spree; a large open-source Rails app that we can use as an example.

Spree has a [`Spree::Base`](https://github.com/spree/spree/blob/master/core/app/models/spree/base.rb) which they use to pump methods and `include`s into that many classes might use. Then, when creating a new class such as [`Spree::Asset`](https://github.com/spree/spree/blob/master/core/app/models/spree/asset.rb), they will inherit from `Spree::Base` instead of `ActiveRecord::Base`.

Except, the issue is that the `param_key` to that object is different than when you extend from a module.

To give you an example:

``` ruby
module Blog
  def self.use_relative_model_naming
    true
  end

  class Post < ActiveRecord::Base
  end
end

p Blog::Post.model_name.param_key
# => "post"

module Asset
  class Base < ActiveRecord::Base
    def self.use_relative_model_naming
      true
    end
  end
end

module Asset
  class Image < Asset::Base
  end
end

p Asset::Image.model_name.param_key
# => "asset_image"
```

As you can see in the example above, there's a difference. A work-around is as so:

``` ruby
module Asset
  def self.use_relative_model_naming
    true
  end

  class Base < ActiveRecord::Base
  end
end
```

This can be done, but its not as elegant. I think Rails should be able to expect if a class is using it in a child rather than the parent module.

I supplied two new test models into Active Support to highlight the extending from different classes. Specs were made to ensure they followed the exact same 8 tests that the others used. Then the fix was made to use the classes namespace if the current class has `#use_relative_model_naming` returning `true`.

### CODE REVIEW

**Comment by BenMorganIO** [comment]
bump

**Comment by BenMorganIO** [comment]
gentle bump

**Comment by BenMorganIO** [comment]
bump

**Comment by BenMorganIO** [comment]
slight bump

**Comment by rafaelfranca** [comment]
R?

**Comment by rafaelfranca** [comment]
R? @rafaelfranca

**Comment by arthurnn** [comment]
@BenMorganIO can you stop bumping this? Rafael will take his time to review it, I am sure he is busy with other important things. So, please, wait it out.
